pydantic_graph/pydantic_graph/nodes.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import annotations as _annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, is_dataclass
+from functools import cache
+from typing import TYPE_CHECKING, Any, ClassVar, Generic, get_origin, get_type_hints
+
+from typing_extensions import Never, TypeVar
+
+from . import _utils, exceptions
+
+if TYPE_CHECKING:
+    from .state import StateT
+else:
+    StateT = TypeVar('StateT', default=None)
+
+__all__ = 'GraphContext', 'BaseNode', 'End', 'Edge', 'NodeDef', 'RunEndT', 'NodeRunEndT'
+
+RunEndT = TypeVar('RunEndT', default=None)
+"""Type variable for the return type of a graph [`run`][pydantic_graph.graph.Graph.run]."""
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
+"""Type variable for the return type of a node [`run`][pydantic_graph.nodes.BaseNode.run]."""
+
+
+@dataclass
+class GraphContext(Generic[StateT]):
```

Comment:
```suggestion
class GraphRunContext(Generic[StateT]):
```
I'm assuming this only applies to graph runs? If so, I think calling it this would make that more clear. Take it or leave it.